common/src/IO/MapReader.cpp

```diff
@@ -79,56 +81,55 @@ namespace TrenchBroom {
             m_worldBounds = worldBounds;
             parseBrushFaces(format, status);
         }
 
+        // implement MapParser interface
+
         void MapReader::onFormatSet(const Model::MapFormat format) {
             m_factory = &initialize(format);
         }
 
-        void MapReader::onBeginEntity(const size_t line, const std::vector<Model::EntityAttribute>& attributes, const ExtraAttributes& extraAttributes, ParserStatus& status) {
-            const EntityType type = entityType(attributes);
-            switch (type) {
-                case EntityType_Layer:
-                    createLayer(line, attributes, extraAttributes, status);
-                    break;
-                case EntityType_Group:
-                    createGroup(line, attributes, extraAttributes, status);
-                    break;
-                case EntityType_Worldspawn:
-                    m_brushParent = onWorldspawn(attributes, extraAttributes, status);
-                    break;
-                case EntityType_Default:
-                    createEntity(line, attributes, extraAttributes, status);
-                    break;
-            }
+        void MapReader::onBeginEntity(const size_t /* line */, const std::vector<Model::EntityAttribute>& attributes, const ExtraAttributes& extraAttributes, ParserStatus& /* status */) {
+            const size_t brushesBegin = m_brushInfos.size();
+
+            m_entityInfos.push_back(EntityInfo{0, 0, attributes, extraAttributes, brushesBegin, brushesBegin});
         }
 
-        void MapReader::onEndEntity(const size_t startLine, const size_t lineCount, ParserStatus& status) {
-            if (m_currentNode != nullptr)
-                setFilePosition(m_currentNode, startLine, lineCount);
-            else
-                onWorldspawnFilePosition(startLine, lineCount, status);
-            m_currentNode = nullptr;
-            m_brushParent = nullptr;
+        void MapReader::onEndEntity(const size_t startLine, const size_t lineCount, ParserStatus& /* status */) {
+            EntityInfo& entity = m_entityInfos.back();
+            entity.startLine = startLine;
+            entity.lineCount = lineCount;
         }
 
         void MapReader::onBeginBrush(const size_t /* line */, ParserStatus& /* status */) {
-            assert(m_faces.empty());
+            m_brushInfos.push_back(BrushInfo{{}, 0, 0, {}});
         }
 
-        void MapReader::onEndBrush(const size_t startLine, const size_t lineCount, const ExtraAttributes& extraAttributes, ParserStatus& status) {
-            createBrush(startLine, lineCount, extraAttributes, status);
+        void MapReader::onEndBrush(const size_t startLine, const size_t lineCount, const ExtraAttributes& extraAttributes, ParserStatus& /* status */) {
+            BrushInfo& brush = m_brushInfos.back();
+            brush.startLine = startLine;
+            brush.lineCount = lineCount;
+            brush.extraAttributes = extraAttributes;
```

Comment:
I suppose the extra attributes should be moved (and passed by value).

Reply:
That can be left for another PR, it would have a cascade of changes to MapParser etc.